File: src/main/java/config/ConfigManager.java
package config;

import java.io.BufferedReader;
import java.io.BufferedWriter;
import java.io.File;
import java.io.FileReader;
import java.io.FileWriter;
import java.io.IOException;
import util.ErrorLogger;

/**
 * Manages configuration file I/O operations
 */
public class ConfigManager {

    private static final String CONFIG_DIR_NAME = "config";
    private static final String ROLL_COST_FILE = "roll_cost.txt";
    private static final String TIER_MODIFIERS_FILE = "tier_modifiers.txt";
    private static final String OPTIMAL_RANGE_MODIFIER_FILE = "optimal_range_modifier.txt";

    /**
     * Private constructor to prevent instantiation of utility class
     */
    private ConfigManager() {
        // Utility class - no instantiation
    }

    /**
     * Validates a filename to prevent path traversal attacks
     * 
     * @param filename The filename to validate
     * @return true if the filename is safe, false otherwise
     */
    private static boolean isValidFilename(String filename) {
        if (filename == null || filename.isEmpty()) {
            return false;
        }

        // Check for path traversal sequences
        if (filename.contains("..") || filename.contains("/") || filename.contains("\\")) {
            return false;
        }

        // Check for null bytes (potential injection)
        if (filename.contains("\0")) {
            return false;
        }

        // Check for Windows reserved characters
        String reservedChars = "<>:\"|?*";
        for (char c : reservedChars.toCharArray()) {
            if (filename.indexOf(c) >= 0) {
                return false;
            }
        }

        return true;
    }

    /**
     * Safely resolves a file path within the config directory, preventing path traversal
     * 
     * @param configDir The base config directory
     * @param filename The filename to resolve
     * @return A validated File object within the config directory, or null if validation fails
     */
    private static File resolveConfigFile(File configDir, String filename) {
        // Validate filename
        if (!isValidFilename(filename)) {
            ErrorLogger.logError(
                    "Invalid filename detected (potential path traversal): " + filename,
                    new SecurityException("Path traversal attempt blocked"));
            return null;
        }

        try {
            File file = new File(configDir, filename);

            // Normalize the path to resolve any remaining issues
            File canonicalFile = file.getCanonicalFile();
            File canonicalConfigDir = configDir.getCanonicalFile();

            // Ensure the resolved file is within the config directory
            String canonicalFilePath = canonicalFile.getPath();
            String canonicalConfigPath = canonicalConfigDir.getPath();

            // Check if the file path starts with the config directory path
            if (!canonicalFilePath.startsWith(canonicalConfigPath + File.separator)
                    && !canonicalFilePath.equals(canonicalConfigPath)) {
                ErrorLogger.logError(
                        "Path traversal detected: resolved path outside config directory",
                        new SecurityException("Path traversal attempt blocked"));
                return null;
            }

            return canonicalFile;
        } catch (IOException e) {
            ErrorLogger.logError("Error resolving config file path: " + filename, e);
            return null;
        }
    }

    /**
     * Converts a URL location to a File, handling URISyntaxException
     * 
     * @param location The URL location to convert
     * @return The File representation of the URL location
     */
    private static File urlToFile(java.net.URL location) {
        try {
            return new File(location.toURI());
        } catch (java.net.URISyntaxException e) {
            ErrorLogger.logError("Failed to convert URL to URI for config directory", e);
            // If URI conversion fails, try getting path directly
            String path = location.getPath();
            // Handle Windows paths: remove leading / if present (e.g., /C:/path -> C:/path)
            if (System.getProperty("os.name").toLowerCase().startsWith("windows")
                    && path.length() > 2 && path.startsWith("/") && path.charAt(2) == ':') {
                path = path.substring(1);
            }
            return new File(path);
        }
    }

    /**
     * Gets base directory from JAR file (directory if IDE, parent if JAR)
     */
    private static File getBaseDirectoryFromJarFile(File jarFile) {
        if (jarFile.isDirectory()) {
            return jarFile;
        }
        return jarFile.getParentFile();
    }

    /**
     * Ensures config directory exists, creating it if necessary
     */
    private static File ensureConfigDirectoryExists(File baseDir, String configDirName) {
        File configDir = new File(baseDir, configDirName);
        if (!configDir.exists()) {
            configDir.mkdirs();
        }
        return configDir;
    }

    /**
     * Tries to get config directory from JAR location
     * 
     * @return Config directory if successful, null otherwise
     */
    private static File tryGetConfigDirectoryFromJar() {
        try {
            java.net.URL location =
                    ConfigManager.class.getProtectionDomain().getCodeSource().getLocation();
            if (location == null) {
                return null;
            }

            File jarFile = urlToFile(location);
            File baseDir = getBaseDirectoryFromJarFile(jarFile);

            if (baseDir != null && baseDir.exists()) {
                return ensureConfigDirectoryExists(baseDir, CONFIG_DIR_NAME);
            }
        } catch (Exception e) {
            ErrorLogger.logError("Error getting config directory from JAR location", e);
            // Fallback to current directory
        }
        return null;
    }

    /**
     * Gets the config directory path (relative to JAR location or current directory)
     */
    private static File getConfigDirectory() {
        File configDir = tryGetConfigDirectoryFromJar();
        if (configDir != null) {
            return configDir;
        }

        // Fallback: use current directory
        return ensureConfigDirectoryExists(new File("."), CONFIG_DIR_NAME);
    }

    /**
     * Gets the roll cost from config file
     * 
     * @return Roll cost as double, or 0.0 if file doesn't exist or is invalid
     */
    public static double getRollCost() {
        File configDir = getConfigDirectory();
        File costFile = resolveConfigFile(configDir, ROLL_COST_FILE);

        if (costFile == null) {
            ErrorLogger.logError("Failed to resolve roll cost file path due to validation failure",
                    new SecurityException("Path validation failed"));
            return 0.0;
        }

        if (!costFile.exists()) {
            // Create default file with 0
            saveRollCost(0.0);
            return 0.0;
        }

        try (BufferedReader reader = new BufferedReader(new FileReader(costFile))) {
            String line = reader.readLine();
            if (line != null) {
                line = line.trim();
                return Double.parseDouble(line);
            }
        } catch (IOException | NumberFormatException e) {
            ErrorLogger.logError(
                    "Error reading roll cost from config file: " + costFile.getAbsolutePath(), e);
            // Return default value on error
        }

        return 0.0;
    }

    /**
     * Saves the roll cost to config file
     * 
     * @param cost The cost to save
     */
    public static void saveRollCost(double cost) {
        File configDir = getConfigDirectory();
        File costFile = resolveConfigFile(configDir, ROLL_COST_FILE);

        if (costFile == null) {
            ErrorLogger.logError("Failed to resolve roll cost file path due to validation failure",
                    new SecurityException("Path validation failed"));
            return;
        }

        try (BufferedWriter writer = new BufferedWriter(new FileWriter(costFile))) {
            writer.write(String.valueOf(cost));
        } catch (IOException e) {
            ErrorLogger.logError(
                    "Error saving roll cost to config file: " + costFile.getAbsolutePath(), e);
            // Config saving failure is logged but not critical
        }
    }

    /**
     * Gets the tier modifiers file path
     * 
     * @return The tier modifiers file, or null if path validation fails
     */
    public static File getTierModifiersFile() {
        File configDir = getConfigDirectory();
        File file = resolveConfigFile(configDir, TIER_MODIFIERS_FILE);

        if (file == null) {
            ErrorLogger.logError(
                    "Failed to resolve tier modifiers file path due to validation failure",
                    new SecurityException("Path validation failed"));
        }

        return file;
    }

    /**
     * Gets the optimal range modifier file path
     * 
     * @return The optimal range modifier file, or null if path validation fails
     */
    public static File getOptimalRangeModifiersFile() {
        File configDir = getConfigDirectory();
        File file = resolveConfigFile(configDir, OPTIMAL_RANGE_MODIFIER_FILE);

        if (file == null) {
            ErrorLogger.logError(
                    "Failed to resolve optimal range modifier file path due to validation failure",
                    new SecurityException("Path validation failed"));
        }

        return file;
    }
}

File: src/lib/config/configManager.ts
import { readTextFile, writeTextFile, exists, BaseDirectory } from '@tauri-apps/plugin-fs';
import type { TierModifiers } from '@/types';

const CONFIG_DIR = 'config';
const ROLL_COST_FILE = 'roll_cost.txt';
const TIER_MODIFIERS_FILE = 'tier_modifiers.txt';
const OPTIMAL_RANGE_MODIFIER_FILE = 'optimal_range_modifier.txt';

async function ensureConfigDir(): Promise<void> {
  // Directory will be created automatically by writeTextFile if it doesn't exist
  // No explicit directory creation needed
}

function getConfigPath(filename: string): string {
  return `${CONFIG_DIR}/${filename}`;
}

export async function getRollCost(): Promise<number> {
  await ensureConfigDir();
  const filePath = getConfigPath(ROLL_COST_FILE);
  
  try {
    const existsFile = await exists(filePath, { baseDir: BaseDirectory.AppData });
    if (!existsFile) {
      return 0;
    }
    
    const content = await readTextFile(filePath, { baseDir: BaseDirectory.AppData });
    const trimmed = content.trim();
    if (!trimmed || trimmed.startsWith('#')) {
      return 0;
    }
    
    const value = Number.parseFloat(trimmed);
    return Number.isNaN(value) || value < 0 ? 0 : value;
  } catch (error) {
    console.error('Error reading roll cost:', error);
    return 0;
  }
}

export async function saveRollCost(cost: number): Promise<void> {
  await ensureConfigDir();
  const filePath = getConfigPath(ROLL_COST_FILE);
  
  try {
    await writeTextFile(filePath, cost.toString(), { baseDir: BaseDirectory.AppData });
  } catch (error) {
    console.error('Error saving roll cost:', error);
    throw error;
  }
}

const DEFAULT_TIER_MODIFIERS: TierModifiers = {
  S: 2,
  A: 1.8,
  B: 1.6,
  C: 1.4,
  D: 1.2,
  E: 1,
  F: 0.8,
};

export async function loadTierModifiers(): Promise<TierModifiers> {
  await ensureConfigDir();
  const filePath = getConfigPath(TIER_MODIFIERS_FILE);
  
  const modifiers: TierModifiers = { ...DEFAULT_TIER_MODIFIERS };
  
  try {
    const existsFile = await exists(filePath, { baseDir: BaseDirectory.AppData });
    if (!existsFile) {
      await saveTierModifiers(modifiers);
      return modifiers;
    }
    
    const content = await readTextFile(filePath, { baseDir: BaseDirectory.AppData });
    const lines = content.split('\n');
    
    for (const line of lines) {
      const trimmed = line.trim();
      if (!trimmed || trimmed.startsWith('#')) {
        continue;
      }
      
      const [tier, value] = trimmed.split('=');
      if (tier && value) {
        const tierKey = tier.trim().toUpperCase() as keyof TierModifiers;
        const numValue = Number.parseFloat(value.trim());
        if (!Number.isNaN(numValue) && tierKey in DEFAULT_TIER_MODIFIERS) {
          modifiers[tierKey] = numValue;
        }
      }
    }
  } catch (error) {
    console.error('Error loading tier modifiers:', error);
  }
  
  return modifiers;
}

export async function saveTierModifiers(modifiers: TierModifiers): Promise<void> {
  await ensureConfigDir();
  const filePath = getConfigPath(TIER_MODIFIERS_FILE);
  
  try {
    const lines: string[] = [];
    const tiers: Array<keyof TierModifiers> = ['S', 'A', 'B', 'C', 'D', 'E', 'F'];
    for (const tier of tiers) {
      lines.push(`${tier}=${modifiers[tier]}`);
    }
    await writeTextFile(filePath, lines.join('\n'), { baseDir: BaseDirectory.AppData });
  } catch (error) {
    console.error('Error saving tier modifiers:', error);
    throw error;
  }
}

export async function loadOptimalRangeModifier(): Promise<number> {
  await ensureConfigDir();
  const filePath = getConfigPath(OPTIMAL_RANGE_MODIFIER_FILE);
  const DEFAULT_MODIFIER = 1;
  
  try {
    const existsFile = await exists(filePath, { baseDir: BaseDirectory.AppData });
    if (!existsFile) {
      await saveOptimalRangeModifier(DEFAULT_MODIFIER);
      return DEFAULT_MODIFIER;
    }
    
    const content = await readTextFile(filePath, { baseDir: BaseDirectory.AppData });
    const trimmed = content.trim().split('\n')[0]?.trim();
    if (!trimmed || trimmed.startsWith('#')) {
      return DEFAULT_MODIFIER;
    }
    
    const value = Number.parseFloat(trimmed);
    return Number.isNaN(value) || value < 0 ? DEFAULT_MODIFIER : value;
  } catch (error) {
    console.error('Error loading optimal range modifier:', error);
    return DEFAULT_MODIFIER;
  }
}

export async function saveOptimalRangeModifier(modifier: number): Promise<void> {
  await ensureConfigDir();
  const filePath = getConfigPath(OPTIMAL_RANGE_MODIFIER_FILE);
  
  try {
    await writeTextFile(filePath, modifier.toString(), { baseDir: BaseDirectory.AppData });
  } catch (error) {
    console.error('Error saving optimal range modifier:', error);
    throw error;
  }
}

export function getModifierForTier(
  tier: string,
  modifiers: TierModifiers,
): number {
  const tierKey = tier.toUpperCase() as keyof TierModifiers;
  return modifiers[tierKey] ?? modifiers.F;
}

export function calculateSellPrice(
  rollCost: number,
  tier: string,
  baseM3Pct: number,
  tierModifiers: TierModifiers,
  optimalRangeModifier: number,
): number {
  if (rollCost <= 0) {
    return 0;
  }
  
  // Remove "+" suffix if present for tier modifier lookup
  const tierForModifier = tier.endsWith('+') ? tier.slice(0, -1) : tier;
  const tierModifier = getModifierForTier(tierForModifier, tierModifiers);
  
  // Check if tier has "+" suffix (optimal range increased) and apply modifier
  const optimalModifier = tier.endsWith('+') ? optimalRangeModifier : 1;
  
  // Formula: cost * tier_modifier * (100% + baseM3Pct%) * optimal_range_modifier (if applicable)
  // baseM3Pct is already a percentage, so convert to multiplier: 1 + (baseM3Pct / 100)
  const m3Multiplier = 1 + baseM3Pct / 100;
  return rollCost * tierModifier * m3Multiplier * optimalModifier;
}
